### ISSUE TITLE
[libFirebaseCppApp-10_1_1.so] firebase::firestore::FieldValueInternal::FieldValueInternal(long)

### ISSUE THREAD

**Issue by nitzanwilnai** [firebase/firebase-unity-sdk#586]
Hi, 
About 15% of our users seem to have Firebase related crashes, it seems when saving to FireStore.

After uploading symbols to Google Play, we managed to get the following crash data in the Google Play Console:
`[libFirebaseCppApp-10_1_1.so] firebase::firestore::FieldValueInternal::FieldValueInternal(long)`

`[libFirebaseCppApp-10_1_1.so] firebase::firestore::jni::Global<firebase::firestore::jni::Object>::Global(firebase::firestore::jni::Object const&)`

`[libFirebaseCppApp-10_1_1.so] firebase::firestore::FieldValue::FieldValue(firebase::firestore::FieldValue const&)`

Some additional information:
After adding symbols to Firebase, we have the following crashes on crashlytics:
```
libil2cpp.so FirestoreCppPINVOKE.cs ConvertVectorToFieldValue

libil2cpp.so FirestoreCppPINVOKE.cs FieldValueProxy_Integer

libil2cpp.so FirestoreCppPINVOKE.cs FieldValueVector_PushBack

```
All three crashes are happening when saving our UserData to FireStore, and they are each happening to different users. So a user who experiences one does not experience the others.

Any ideas?

Thanks

**Comment by nitzanwilnai** [comment]
The code to send data to Firestore is very simple:

```
        public void SendUserDataToFireStore(UserData userData)
        {
            FirebaseTimeout = Constants.FIREBASE_TIMEOUT;

            FirebaseFirestore db = FirebaseFirestore.DefaultInstance;

            if (PlayerPrefs.GetInt(FirebaseUser.UserId) == 0)
                PlayerPrefs.SetInt(FirebaseUser.UserId, 1);

            DocumentReference docRef = db.Collection(Constants.UserDataCollection).Document(FirebaseUser.UserId);
            docRef.SetAsync(userData).ContinueWith(task =>
            {
                if (task.IsCompleted)
                {

                    Debug.Log("UserData DocRef ID " + docRef.Id);

                }
                else
                {
                    Debug.Log("SendUserDataToFireStore ERROR ADDING DATA TO FIRESTORE!");
                }
            });
        }

```

**Comment by cherylEnkidu** [comment]
Hi,

Thank you for filling up the ticket! The team is not available for support during the holiday break and we will get back to you in the new year. Happy holiday!

**Comment by nitzanwilnai** [comment]
Some additional information:
Unity 2020.3.41f1
Firebase Unity SDK 10.1.1

Crashlytics reports - 

<img width="1258" alt="Screen Shot 2023-01-02 at 8 56 26 AM" src="https://user-images.githubusercontent.com/1131364/210202302-61be4f09-dbba-4ab2-9b87-9fb36163499e.png">

<img width="1266" alt="Screen Shot 2023-01-02 at 8 56 46 AM" src="https://user-images.githubusercontent.com/1131364/210202289-d553ebba-1eca-4741-8a17-41b7d63653e5.png">

<img width="1252" alt="Screen Shot 2023-01-02 at 8 57 02 AM" src="https://user-images.githubusercontent.com/1131364/210202319-48f24c14-d3ef-4b73-bcb1-25e401285e51.png">

**Comment by cherylEnkidu** [comment]
Hi @nitzanwilnai ,

Could you provide the crash report in detail? For example what is the error message when crashes?

**Comment by nitzanwilnai** [comment]
I am not sure what you are asking. I included all the information above. The messages in Google Play and Firebase and the stack traces. There is no error message, only the crash report. We know that on the user device in production the game simply exits.

Note that every single user successfully saved their data to Firestore multiple times. Most had 20+ minute sessions where they saved the data dozens of times without issue before the crash happened.

If it makes it easier I can combine all the info here:

Firebase:
<img width="1297" alt="Screen Shot 2023-01-03 at 8 16 32 AM" src="https://user-images.githubusercontent.com/1131364/210308294-e5a1eecd-d8ef-47c8-b1e3-a134f8a52188.png">
<img width="1252" alt="Screen Shot 2023-01-02 at 8 57 02 AM" src="https://user-images.githubusercontent.com/1131364/210308528-0abbff76-f10e-4279-95be-19f9cfe60597.png">
<img width="1266" alt="Screen Shot 2023-01-02 at 8 56 46 AM" src="https://user-images.githubusercontent.com/1131364/210308540-72a05fe7-7517-4d86-bd3f-fb14968238a7.png">
<img width="1258" alt="Screen Shot 2023-01-02 at 8 56 26 AM" src="https://user-images.githubusercontent.com/1131364/210308547-a4a24788-9df5-4995-b00c-f33eb856db4e.png">


Google Play:
<img width="1307" alt="Screen Shot 2023-01-03 at 8 22 51 AM" src="https://user-images.githubusercontent.com/1131364/210308694-bba92925-bdd8-46e8-aeb6-e09cea5d36dc.png">
*** *** *** *** *** *** *** *** *** *** *** *** *** *** *** ***
pid: 0, tid: 3325 >>> com.fourxp.mergeup <<<

backtrace:
  #00  pc 0x000000000004e178  /apex/com.android.runtime/lib64/bionic/libc.so (abort+168)
  #01  pc 0x00000000005527c4  /apex/com.android.art/lib64/libart.so (art::Runtime::Abort(char const*)+2260)
  #02  pc 0x0000000000013990  /system/lib64/libbase.so
  #03  pc 0x0000000000012fb4  /system/lib64/libbase.so (android::base::LogMessage::~LogMessage()+320)
  #04  pc 0x0000000000384138  /apex/com.android.art/lib64/libart.so (art::JavaVMExt::AddGlobalRef(art::Thread*, art::ObjPtr<art::mirror::Object>)+268)
  #05  pc 0x0000000000396748  /apex/com.android.art/lib64/libart.so (art::JNI<false>::NewGlobalRef(_JNIEnv*, _jobject*)+632)
  #06  pc 0x00000000002407f4  /data/app/~~YKjx_Wji-qMGXyEI-Xcwrg==/com.fourxp.mergeup-3xha8t9Ls_oyeKHc0thx6A==/lib/arm64/libFirebaseCppApp-10_1_1.so (firebase::firestore::FieldValue::FieldValue(firebase::firestore::FieldValue const&)+100)
  #07  pc 0x0000000000209830  /data/app/~~YKjx_Wji-qMGXyEI-Xcwrg==/com.fourxp.mergeup-3xha8t9Ls_oyeKHc0thx6A==/lib/arm64/libFirebaseCppApp-10_1_1.so (std::__ndk1::vector<firebase::firestore::FieldValue, std::__ndk1::allocator<firebase::firestore::FieldValue> >::vector(std::__ndk1::vector<firebase::firestore::FieldValue, std::__ndk1::allocator<firebase::firestore::FieldValue> > const&)+96)
  #08  pc 0x0000000000205ab8  /data/app/~~YKjx_Wji-qMGXyEI-Xcwrg==/com.fourxp.mergeup-3xha8t9Ls_oyeKHc0thx6A==/lib/arm64/libFirebaseCppApp-10_1_1.so (Firebase_Firestore_CSharp_ConvertVectorToFieldValue+56)
  #09  pc 0x00000000005b8c8c  /data/app/~~YKjx_Wji-qMGXyEI-Xcwrg==/com.fourxp.mergeup-3xha8t9Ls_oyeKHc0thx6A==/lib/arm64/libil2cpp.so (FirestoreCppPINVOKE_ConvertVectorToFieldValue_mE35676BFE78C31EC779CD5E78F6F7737EF932136+30757)
  #10  pc 0x00000000005a8ed8  /data/app/~~YKjx_Wji-qMGXyEI-Xcwrg==/com.fourxp.mergeup-3xha8t9Ls_oyeKHc0thx6A==/lib/arm64/libil2cpp.so (FirestoreCpp_ConvertVectorToFieldValue_mAB106034884875826A1E251F2CF1DA172EA95EBA+27390)
  #11  pc 0x00000000005b12c0  /data/app/~~YKjx_Wji-qMGXyEI-Xcwrg==/com.fourxp.mergeup-3xha8t9Ls_oyeKHc0thx6A==/lib/arm64/libil2cpp.so (EnumerableConverterBase_Serialize_m40FF859EA17BAEE65D7765ADF75C903F1DE77792+21422)
  #12  pc 0x00000000005a4e74  /data/app/~~YKjx_Wji-qMGXyEI-Xcwrg==/com.fourxp.mergeup-3xha8t9Ls_oyeKHc0thx6A==/lib/arm64/libil2cpp.so (AttributedTypeConverter_SerializeMap_m2AA6C554C645BF93A347C40E5D0F00F43FC8C10B+14823)
  #13  pc 0x00000000005bf4bc  /data/app/~~YKjx_Wji-qMGXyEI-Xcwrg==/com.fourxp.mergeup-3xha8t9Ls_oyeKHc0thx6A==/lib/arm64/libil2cpp.so (MapConverterBase_Serialize_mF24CAD26C3A877991AE6732907C94C39078FD206+44)
  #14  pc 0x00000000005a6604  /data/app/~~YKjx_Wji-qMGXyEI-Xcwrg==/com.fourxp.mergeup-3xha8t9Ls_oyeKHc0thx6A==/lib/arm64/libil2cpp.so (DocumentReference_SetAsync_mA0B8D6AB674F0ADF47F693B58D5309B8B84E7451+18859)
  #15  pc 0x0000000000ae7160  /data/app/~~YKjx_Wji-qMGXyEI-Xcwrg==/com.fourxp.mergeup-3xha8t9Ls_oyeKHc0thx6A==/lib/arm64/libil2cpp.so (FirebaseManager_SendBusinessDataToFireStore_m6452EF303E7E3AA9A6FA43904C0B5084528B73A3+39835)
  #16  pc 0x00000000007d1eb8  /data/app/~~YKjx_Wji-qMGXyEI-Xcwrg==/com.fourxp.mergeup-3xha8t9Ls_oyeKHc0thx6A==/lib/arm64/libil2cpp.so (Action_1_Invoke_m587509C88BB83721D7918D89DF07606BB752D744_gshared+47)
  #17  pc 0x0000000000ac0af0  /data/app/~~YKjx_Wji-qMGXyEI-Xcwrg==/com.fourxp.mergeup-3xha8t9Ls_oyeKHc0thx6A==/lib/arm64/libil2cpp.so (U3CU3Ec__DisplayClass0_1_U3CContinueWithOnMainThreadU3Eb__1_m669E73DC507424634BD00848B6A61F918DCF75DA+921)
  #18  pc 0x00000000007ab720  /data/app/~~YKjx_Wji-qMGXyEI-Xcwrg==/com.fourxp.mergeup-3xha8t9Ls_oyeKHc0thx6A==/lib/arm64/libil2cpp.so (Func_1_Invoke_mD0192F3F912E371387CD7B196DF85569553486B4_gshared+47)
  #19  pc 0x00000000007ce6a4  /data/app/~~YKjx_Wji-qMGXyEI-Xcwrg==/com.fourxp.mergeup-3xha8t9Ls_oyeKHc0thx6A==/lib/arm64/libil2cpp.so (U3CU3Ec__DisplayClass5_0_1_U3CRunAsyncU3Eb__0_m0AAB5C27727D26D190EB040F6E10869615D38C75_gshared+9925)
  #20  pc 0x0000000000745e68  /data/app/~~YKjx_Wji-qMGXyEI-Xcwrg==/com.fourxp.mergeup-3xha8t9Ls_oyeKHc0thx6A==/lib/arm64/libil2cpp.so (Action_Invoke_m3FFA5BE3D64F0FF8E1E1CB6F953913FADB5EB89E+47)
  #21  pc 0x0000000000a97688  /data/app/~~YKjx_Wji-qMGXyEI-Xcwrg==/com.fourxp.mergeup-3xha8t9Ls_oyeKHc0thx6A==/lib/arm64/libil2cpp.so (ExceptionAggregator_Wrap_m286D2940E6E4C43E3C9B5E23A249A63B731390B4+5296)
  #22  pc 0x0000000000a97580  /data/app/~~YKjx_Wji-qMGXyEI-Xcwrg==/com.fourxp.mergeup-3xha8t9Ls_oyeKHc0thx6A==/lib/arm64/libil2cpp.so (Dispatcher_PollJobs_m4DD47E8B63F33D975BC1A7D670A509F597F8E9E0+4915)
  #23  pc 0x0000000000a98a0c  /data/app/~~YKjx_Wji-qMGXyEI-Xcwrg==/com.fourxp.mergeup-3xha8t9Ls_oyeKHc0thx6A==/lib/arm64/libil2cpp.so (FirebaseHandler_Update_m054C867D529EEC80853232EEAFBCAD18CC287817+6523)
  #24  pc 0x000000000029b7bc  /data/app/~~YKjx_Wji-qMGXyEI-Xcwrg==/com.fourxp.mergeup-3xha8t9Ls_oyeKHc0thx6A==/lib/arm64/libil2cpp.so (RuntimeInvoker_TrueVoid_t700C6383A2A510C2CF4DD86DABD5CA9FF70ADAC5(void (*)(), MethodInfo const*, void*, void**)+44228)
  #25  pc 0x000000000035d4b8  /data/app/~~YKjx_Wji-qMGXyEI-Xcwrg==/com.fourxp.mergeup-3xha8t9Ls_oyeKHc0thx6A==/lib/arm64/libil2cpp.so (il2cpp::vm::Runtime::Invoke(MethodInfo const*, void*, void**, Il2CppException**)+556)
  #26  pc 0x00000000002bebc0  /data/app/~~YKjx_Wji-qMGXyEI-Xcwrg==/com.fourxp.mergeup-3xha8t9Ls_oyeKHc0thx6A==/lib/arm64/libunity.so (scripting_method_invoke(ScriptingMethodPtr, ScriptingObjectPtr, ScriptingArguments&, ScriptingExceptionPtr*, bool))
  #27  pc 0x00000000002c21fc  /data/app/~~YKjx_Wji-qMGXyEI-Xcwrg==/com.fourxp.mergeup-3xha8t9Ls_oyeKHc0thx6A==/lib/arm64/libunity.so (ScriptingInvocation::Invoke(ScriptingExceptionPtr*, bool))
  #28  pc 0x00000000002d76a4  /data/app/~~YKjx_Wji-qMGXyEI-Xcwrg==/com.fourxp.mergeup-3xha8t9Ls_oyeKHc0thx6A==/lib/arm64/libunity.so (MonoBehaviour::CallUpdateMethod(int))
  #29  pc 0x00000000001ad0cc  /data/app/~~YKjx_Wji-qMGXyEI-Xcwrg==/com.fourxp.mergeup-3xha8t9Ls_oyeKHc0thx6A==/lib/arm64/libunity.so (void BaseBehaviourManager::CommonUpdate<BehaviourManager>())
  #30  pc 0x000000000022456c  /data/app/~~YKjx_Wji-qMGXyEI-Xcwrg==/com.fourxp.mergeup-3xha8t9Ls_oyeKHc0thx6A==/lib/arm64/libunity.so (ExecutePlayerLoop(NativePlayerLoopSystem*))
  #31  pc 0x00000000002245a0  /data/app/~~YKjx_Wji-qMGXyEI-Xcwrg==/com.fourxp.mergeup-3xha8t9Ls_oyeKHc0thx6A==/lib/arm64/libunity.so (ExecutePlayerLoop(NativePlayerLoopSystem*))
  #32  pc 0x00000000002247e4  /data/app/~~YKjx_Wji-qMGXyEI-Xcwrg==/com.fourxp.mergeup-3xha8t9Ls_oyeKHc0thx6A==/lib/arm64/libunity.so (PlayerLoop())
  #33  pc 0x000000000033b608  /data/app/~~YKjx_Wji-qMGXyEI-Xcwrg==/com.fourxp.mergeup-3xha8t9Ls_oyeKHc0thx6A==/lib/arm64/libunity.so (UnityPlayerLoop())
  #34  pc 0x0000000000351158  /data/app/~~YKjx_Wji-qMGXyEI-Xcwrg==/com.fourxp.mergeup-3xha8t9Ls_oyeKHc0thx6A==/lib/arm64/libunity.so (nativeRender(_JNIEnv*, _jobject*))
  #35  pc 0x0000000000041a3c  /data/app/~~YKjx_Wji-qMGXyEI-Xcwrg==/com.fourxp.mergeup-3xha8t9Ls_oyeKHc0thx6A==/oat/arm64/base.odex

<img width="1317" alt="Screen Shot 2023-01-03 at 8 22 42 AM" src="https://user-images.githubusercontent.com/1131364/210308698-4d060070-c9c2-456f-8356-855fd0ab6c18.png">
*** *** *** *** *** *** *** *** *** *** *** *** *** *** *** ***
pid: 0, tid: 14281 >>> com.fourxp.mergeup <<<

backtrace:
  #00  pc 0x0000000000050034  /apex/com.android.runtime/lib64/bionic/libc.so (abort+180)
  #01  pc 0x000000000006504c  /apex/com.android.runtime/lib64/bionic/libc.so (getauxval)
  #02  pc 0x000000000045cad8  /apex/com.android.art/lib64/libart.so (art::JavaVMExt::CheckGlobalRefAllocationTracking())
  #03  pc 0x00000000005cf9b4  /apex/com.android.art/lib64/libart.so (art::JNI<false>::NewGlobalRef(_JNIEnv*, _jobject*)+152)
  #04  pc 0x00000000002407f4  /data/app/~~tvDC2UaGB4QFqJoFIn9NJw==/com.fourxp.mergeup-qKsnOKIVfq95mYM0SMkauw==/lib/arm64/libFirebaseCppApp-10_1_1.so (firebase::firestore::FieldValue::FieldValue(firebase::firestore::FieldValue const&)+100)
  #05  pc 0x000000000020830c  /data/app/~~tvDC2UaGB4QFqJoFIn9NJw==/com.fourxp.mergeup-qKsnOKIVfq95mYM0SMkauw==/lib/arm64/libFirebaseCppApp-10_1_1.so (Firebase_Firestore_CSharp_FieldValueVector_PushBack+40)
  #06  pc 0x00000000005b4bcc  /data/app/~~tvDC2UaGB4QFqJoFIn9NJw==/com.fourxp.mergeup-qKsnOKIVfq95mYM0SMkauw==/lib/arm64/libil2cpp.so (FirestoreCppPINVOKE_FieldValueVector_PushBack_m6F53C48070883074B9BC581A38373E5C2A1C4F5B+30945)
  #07  pc 0x00000000005a8dbc  /data/app/~~tvDC2UaGB4QFqJoFIn9NJw==/com.fourxp.mergeup-qKsnOKIVfq95mYM0SMkauw==/lib/arm64/libil2cpp.so (FieldValueVector_PushBack_mA57371306F100708C468FA543AF22A302D4B0E64+25282)
  #08  pc 0x00000000005a8ba4  /data/app/~~tvDC2UaGB4QFqJoFIn9NJw==/com.fourxp.mergeup-qKsnOKIVfq95mYM0SMkauw==/lib/arm64/libil2cpp.so (ConverterBase_ConvertToProxyArray_m15B85B6AC3B0D8B4B7D9B84D516844EF3FF1E4C4+16730)
  #09  pc 0x00000000005b12c0  /data/app/~~tvDC2UaGB4QFqJoFIn9NJw==/com.fourxp.mergeup-qKsnOKIVfq95mYM0SMkauw==/lib/arm64/libil2cpp.so (EnumerableConverterBase_Serialize_m40FF859EA17BAEE65D7765ADF75C903F1DE77792+21422)
  #10  pc 0x00000000005a4e74  /data/app/~~tvDC2UaGB4QFqJoFIn9NJw==/com.fourxp.mergeup-qKsnOKIVfq95mYM0SMkauw==/lib/arm64/libil2cpp.so (AttributedTypeConverter_SerializeMap_m2AA6C554C645BF93A347C40E5D0F00F43FC8C10B+14823)
  #11  pc 0x00000000005bf4bc  /data/app/~~tvDC2UaGB4QFqJoFIn9NJw==/com.fourxp.mergeup-qKsnOKIVfq95mYM0SMkauw==/lib/arm64/libil2cpp.so (MapConverterBase_Serialize_mF24CAD26C3A877991AE6732907C94C39078FD206+44)
  #12  pc 0x00000000005a6604  /data/app/~~tvDC2UaGB4QFqJoFIn9NJw==/com.fourxp.mergeup-qKsnOKIVfq95mYM0SMkauw==/lib/arm64/libil2cpp.so (DocumentReference_SetAsync_mA0B8D6AB674F0ADF47F693B58D5309B8B84E7451+18859)
  #13  pc 0x0000000000ae7160  /data/app/~~tvDC2UaGB4QFqJoFIn9NJw==/com.fourxp.mergeup-qKsnOKIVfq95mYM0SMkauw==/lib/arm64/libil2cpp.so (FirebaseManager_SendBusinessDataToFireStore_m6452EF303E7E3AA9A6FA43904C0B5084528B73A3+39835)
  #14  pc 0x00000000007d1eb8  /data/app/~~tvDC2UaGB4QFqJoFIn9NJw==/com.fourxp.mergeup-qKsnOKIVfq95mYM0SMkauw==/lib/arm64/libil2cpp.so (Action_1_Invoke_m587509C88BB83721D7918D89DF07606BB752D744_gshared+47)
  #15  pc 0x0000000000ac0af0  /data/app/~~tvDC2UaGB4QFqJoFIn9NJw==/com.fourxp.mergeup-qKsnOKIVfq95mYM0SMkauw==/lib/arm64/libil2cpp.so (U3CU3Ec__DisplayClass0_1_U3CContinueWithOnMainThreadU3Eb__1_m669E73DC507424634BD00848B6A61F918DCF75DA+921)
  #16  pc 0x00000000007ab720  /data/app/~~tvDC2UaGB4QFqJoFIn9NJw==/com.fourxp.mergeup-qKsnOKIVfq95mYM0SMkauw==/lib/arm64/libil2cpp.so (Func_1_Invoke_mD0192F3F912E371387CD7B196DF85569553486B4_gshared+47)
  #17  pc 0x00000000007ce6a4  /data/app/~~tvDC2UaGB4QFqJoFIn9NJw==/com.fourxp.mergeup-qKsnOKIVfq95mYM0SMkauw==/lib/arm64/libil2cpp.so (U3CU3Ec__DisplayClass5_0_1_U3CRunAsyncU3Eb__0_m0AAB5C27727D26D190EB040F6E10869615D38C75_gshared+9925)
  #18  pc 0x0000000000745e68  /data/app/~~tvDC2UaGB4QFqJoFIn9NJw==/com.fourxp.mergeup-qKsnOKIVfq95mYM0SMkauw==/lib/arm64/libil2cpp.so (Action_Invoke_m3FFA5BE3D64F0FF8E1E1CB6F953913FADB5EB89E+47)
  #19  pc 0x0000000000a97688  /data/app/~~tvDC2UaGB4QFqJoFIn9NJw==/com.fourxp.mergeup-qKsnOKIVfq95mYM0SMkauw==/lib/arm64/libil2cpp.so (ExceptionAggregator_Wrap_m286D2940E6E4C43E3C9B5E23A249A63B731390B4+5296)
  #20  pc 0x0000000000a97580  /data/app/~~tvDC2UaGB4QFqJoFIn9NJw==/com.fourxp.mergeup-qKsnOKIVfq95mYM0SMkauw==/lib/arm64/libil2cpp.so (Dispatcher_PollJobs_m4DD47E8B63F33D975BC1A7D670A509F597F8E9E0+4915)
  #21  pc 0x0000000000a98a0c  /data/app/~~tvDC2UaGB4QFqJoFIn9NJw==/com.fourxp.mergeup-qKsnOKIVfq95mYM0SMkauw==/lib/arm64/libil2cpp.so (FirebaseHandler_Update_m054C867D529EEC80853232EEAFBCAD18CC287817+6523)
  #22  pc 0x000000000029b7bc  /data/app/~~tvDC2UaGB4QFqJoFIn9NJw==/com.fourxp.mergeup-qKsnOKIVfq95mYM0SMkauw==/lib/arm64/libil2cpp.so (RuntimeInvoker_TrueVoid_t700C6383A2A510C2CF4DD86DABD5CA9FF70ADAC5(void (*)(), MethodInfo const*, void*, void**)+44228)
  #23  pc 0x000000000035d4b8  /data/app/~~tvDC2UaGB4QFqJoFIn9NJw==/com.fourxp.mergeup-qKsnOKIVfq95mYM0SMkauw==/lib/arm64/libil2cpp.so (il2cpp::vm::Runtime::Invoke(MethodInfo const*, void*, void**, Il2CppException**)+556)
  #24  pc 0x00000000002bebc0  /data/app/~~tvDC2UaGB4QFqJoFIn9NJw==/com.fourxp.mergeup-qKsnOKIVfq95mYM0SMkauw==/lib/arm64/libunity.so (scripting_method_invoke(ScriptingMethodPtr, ScriptingObjectPtr, ScriptingArguments&, ScriptingExceptionPtr*, bool))
  #25  pc 0x00000000002c21fc  /data/app/~~tvDC2UaGB4QFqJoFIn9NJw==/com.fourxp.mergeup-qKsnOKIVfq95mYM0SMkauw==/lib/arm64/libunity.so (ScriptingInvocation::Invoke(ScriptingExceptionPtr*, bool))
  #26  pc 0x00000000002d76a4  /data/app/~~tvDC2UaGB4QFqJoFIn9NJw==/com.fourxp.mergeup-qKsnOKIVfq95mYM0SMkauw==/lib/arm64/libunity.so (MonoBehaviour::CallUpdateMethod(int))
  #27  pc 0x00000000001ad0cc  /data/app/~~tvDC2UaGB4QFqJoFIn9NJw==/com.fourxp.mergeup-qKsnOKIVfq95mYM0SMkauw==/lib/arm64/libunity.so (void BaseBehaviourManager::CommonUpdate<BehaviourManager>())
  #28  pc 0x000000000022456c  /data/app/~~tvDC2UaGB4QFqJoFIn9NJw==/com.fourxp.mergeup-qKsnOKIVfq95mYM0SMkauw==/lib/arm64/libunity.so (ExecutePlayerLoop(NativePlayerLoopSystem*))
  #29  pc 0x00000000002245a0  /data/app/~~tvDC2UaGB4QFqJoFIn9NJw==/com.fourxp.mergeup-qKsnOKIVfq95mYM0SMkauw==/lib/arm64/libunity.so (ExecutePlayerLoop(NativePlayerLoopSystem*))
  #30  pc 0x00000000002247e4  /data/app/~~tvDC2UaGB4QFqJoFIn9NJw==/com.fourxp.mergeup-qKsnOKIVfq95mYM0SMkauw==/lib/arm64/libunity.so (PlayerLoop())
  #31  pc 0x000000000033b608  /data/app/~~tvDC2UaGB4QFqJoFIn9NJw==/com.fourxp.mergeup-qKsnOKIVfq95mYM0SMkauw==/lib/arm64/libunity.so (UnityPlayerLoop())
  #32  pc 0x0000000000351158  /data/app/~~tvDC2UaGB4QFqJoFIn9NJw==/com.fourxp.mergeup-qKsnOKIVfq95mYM0SMkauw==/lib/arm64/libunity.so (nativeRender(_JNIEnv*, _jobject*))
  #33  pc 0x000000000043dd54  /apex/com.android.art/lib64/libart.so (art_quick_generic_jni_trampoline+148)
  #34  pc 0x000000000020a2b0  /apex/com.android.art/lib64/libart.so (nterp_helper+4016)
  #35  pc 0x0000000002003454  /memfd:jit-cache

<img width="1310" alt="Screen Shot 2023-01-03 at 8 22 35 AM" src="https://user-images.githubusercontent.com/1131364/210308707-623e5e34-4b4a-42da-9909-a0e7e3569b3c.png">
*** *** *** *** *** *** *** *** *** *** *** *** *** *** *** ***
pid: 0, tid: 10167 >>> com.fourxp.mergeup <<<

backtrace:
  #00  pc 0x000000000005ec46  /apex/com.android.runtime/lib/bionic/libc.so (abort+166)
  #01  pc 0x000000000037d651  /apex/com.android.runtime/lib/libart.so (art::Runtime::Abort(char const*)+1600)
  #02  pc 0x000000000000857d  /system/lib/libbase.so (android::base::LogMessage::~LogMessage()+412)
  #03  pc 0x000000000028ac2d  /apex/com.android.runtime/lib/libart.so (art::JavaVMExt::AddGlobalRef(art::Thread*, art::ObjPtr<art::mirror::Object>)+200)
  #04  pc 0x0000000000294e39  /apex/com.android.runtime/lib/libart.so (art::JNI::NewGlobalRef(_JNIEnv*, _jobject*)+456)
  #05  pc 0x0000000000211eb1  /data/app/com.fourxp.mergeup-VlNZ4_OQ3Q7uuUnb5kDQNA==/lib/arm/libFirebaseCppApp-10_1_1.so (firebase::firestore::jni::Global<firebase::firestore::jni::Object>::Global(firebase::firestore::jni::Object const&)+48)
  #06  pc 0x00000000002143c9  /data/app/com.fourxp.mergeup-VlNZ4_OQ3Q7uuUnb5kDQNA==/lib/arm/libFirebaseCppApp-10_1_1.so (firebase::firestore::FieldValueInternal::FieldValueInternal(firebase::firestore::FieldValueInternal const&)+6)
  #07  pc 0x00000000002143a7  /data/app/com.fourxp.mergeup-VlNZ4_OQ3Q7uuUnb5kDQNA==/lib/arm/libFirebaseCppApp-10_1_1.so (firebase::firestore::FieldValue::FieldValue(firebase::firestore::FieldValue const&)+28)
  #08  pc 0x00000000001f64cb  /data/app/com.fourxp.mergeup-VlNZ4_OQ3Q7uuUnb5kDQNA==/lib/arm/libFirebaseCppApp-10_1_1.so (void std::__ndk1::vector<firebase::firestore::FieldValue, std::__ndk1::allocator<firebase::firestore::FieldValue> >::__construct_one_at_end<firebase::firestore::FieldValue const&>(firebase::firestore::FieldValue const&&&)+10)
  #09  pc 0x00000000001f4011  /data/app/com.fourxp.mergeup-VlNZ4_OQ3Q7uuUnb5kDQNA==/lib/arm/libFirebaseCppApp-10_1_1.so (Firebase_Firestore_CSharp_FieldValueVector_PushBack+8)
  #10  pc 0x000000000047e414  /data/app/com.fourxp.mergeup-VlNZ4_OQ3Q7uuUnb5kDQNA==/lib/arm/libil2cpp.so (FirestoreCppPINVOKE_FieldValueVector_PushBack_m6F53C48070883074B9BC581A38373E5C2A1C4F5B+30945)
  #11  pc 0x000000000046ffac  /data/app/com.fourxp.mergeup-VlNZ4_OQ3Q7uuUnb5kDQNA==/lib/arm/libil2cpp.so (FieldValueVector_PushBack_mA57371306F100708C468FA543AF22A302D4B0E64+25282)
  #12  pc 0x000000000046fd1c  /data/app/com.fourxp.mergeup-VlNZ4_OQ3Q7uuUnb5kDQNA==/lib/arm/libil2cpp.so (ConverterBase_ConvertToProxyArray_m15B85B6AC3B0D8B4B7D9B84D516844EF3FF1E4C4+16730)
  #13  pc 0x000000000047a054  /data/app/com.fourxp.mergeup-VlNZ4_OQ3Q7uuUnb5kDQNA==/lib/arm/libil2cpp.so (EnumerableConverterBase_Serialize_m40FF859EA17BAEE65D7765ADF75C903F1DE77792+21422)
  #14  pc 0x000000000046b488  /data/app/com.fourxp.mergeup-VlNZ4_OQ3Q7uuUnb5kDQNA==/lib/arm/libil2cpp.so (AttributedTypeConverter_SerializeMap_m2AA6C554C645BF93A347C40E5D0F00F43FC8C10B+14823)
  #15  pc 0x000000000048aeec  /data/app/com.fourxp.mergeup-VlNZ4_OQ3Q7uuUnb5kDQNA==/lib/arm/libil2cpp.so (MapConverterBase_Serialize_mF24CAD26C3A877991AE6732907C94C39078FD206+44)
  #16  pc 0x000000000046d020  /data/app/com.fourxp.mergeup-VlNZ4_OQ3Q7uuUnb5kDQNA==/lib/arm/libil2cpp.so (DocumentReference_SetAsync_mA0B8D6AB674F0ADF47F693B58D5309B8B84E7451+18859)
  #17  pc 0x0000000000ab0428  /data/app/com.fourxp.mergeup-VlNZ4_OQ3Q7uuUnb5kDQNA==/lib/arm/libil2cpp.so (FirebaseManager_SendBusinessDataToFireStore_m6452EF303E7E3AA9A6FA43904C0B5084528B73A3+39835)
  #18  pc 0x0000000000ac8224  /data/app/com.fourxp.mergeup-VlNZ4_OQ3Q7uuUnb5kDQNA==/lib/arm/libil2cpp.so (Game_SaveBusinessData_mB4107BD702082927FDD9FEF420BDEC0AB8A33114+31062)
  #19  pc 0x0000000000705054  /data/app/com.fourxp.mergeup-VlNZ4_OQ3Q7uuUnb5kDQNA==/lib/arm/libil2cpp.so (Action_1_Invoke_m587509C88BB83721D7918D89DF07606BB752D744_gshared+12812)
  #20  pc 0x0000000000a7df24  /data/app/com.fourxp.mergeup-VlNZ4_OQ3Q7uuUnb5kDQNA==/lib/arm/libil2cpp.so (U3CU3Ec__DisplayClass0_1_U3CContinueWithOnMainThreadU3Eb__1_m669E73DC507424634BD00848B6A61F918DCF75DA+921)
  #21  pc 0x00000000006d517c  /data/app/com.fourxp.mergeup-VlNZ4_OQ3Q7uuUnb5kDQNA==/lib/arm/libil2cpp.so (Func_1_Invoke_mD0192F3F912E371387CD7B196DF85569553486B4_gshared+19691)
  #22  pc 0x000000000070102c  /data/app/com.fourxp.mergeup-VlNZ4_OQ3Q7uuUnb5kDQNA==/lib/arm/libil2cpp.so (U3CU3Ec__DisplayClass5_0_1_U3CRunAsyncU3Eb__0_m0AAB5C27727D26D190EB040F6E10869615D38C75_gshared+9925)
  #23  pc 0x000000000065efc8  /data/app/com.fourxp.mergeup-VlNZ4_OQ3Q7uuUnb5kDQNA==/lib/arm/libil2cpp.so (Action_Invoke_m3FFA5BE3D64F0FF8E1E1CB6F953913FADB5EB89E+21048)
  #24  pc 0x0000000000a4fe18  /data/app/com.fourxp.mergeup-VlNZ4_OQ3Q7uuUnb5kDQNA==/lib/arm/libil2cpp.so (ExceptionAggregator_Wrap_m286D2940E6E4C43E3C9B5E23A249A63B731390B4+5296)
  #25  pc 0x0000000000a4fcd0  /data/app/com.fourxp.mergeup-VlNZ4_OQ3Q7uuUnb5kDQNA==/lib/arm/libil2cpp.so (Dispatcher_PollJobs_m4DD47E8B63F33D975BC1A7D670A509F597F8E9E0+4915)
  #26  pc 0x0000000000a5152c  /data/app/com.fourxp.mergeup-VlNZ4_OQ3Q7uuUnb5kDQNA==/lib/arm/libil2cpp.so (FirebaseHandler_Update_m054C867D529EEC80853232EEAFBCAD18CC287817+6523)
  #27  pc 0x00000000001236e4  /data/app/com.fourxp.mergeup-VlNZ4_OQ3Q7uuUnb5kDQNA==/lib/arm/libil2cpp.so (RuntimeInvoker_TrueVoid_t700C6383A2A510C2CF4DD86DABD5CA9FF70ADAC5(void (*)(), MethodInfo const*, void*, void**)+44228)
  #28  pc 0x00000000001ec848  /data/app/com.fourxp.mergeup-VlNZ4_OQ3Q7uuUnb5kDQNA==/lib/arm/libil2cpp.so (il2cpp::vm::Runtime::Invoke(MethodInfo const*, void*, void**, Il2CppException**)+556)
  #29  pc 0x00000000001d23a1  /data/app/com.fourxp.mergeup-VlNZ4_OQ3Q7uuUnb5kDQNA==/lib/arm/libunity.so (scripting_method_invoke(ScriptingMethodPtr, ScriptingObjectPtr, ScriptingArguments&, ScriptingExceptionPtr*, bool))
  #30  pc 0x00000000001d4aa1  /data/app/com.fourxp.mergeup-VlNZ4_OQ3Q7uuUnb5kDQNA==/lib/arm/libunity.so (ScriptingInvocation::Invoke(ScriptingExceptionPtr*, bool))
  #31  pc 0x00000000001e43e1  /data/app/com.fourxp.mergeup-VlNZ4_OQ3Q7uuUnb5kDQNA==/lib/arm/libunity.so (MonoBehaviour::CallUpdateMethod(int))
  #32  pc 0x0000000000108c1f  /data/app/com.fourxp.mergeup-VlNZ4_OQ3Q7uuUnb5kDQNA==/lib/arm/libunity.so (void BaseBehaviourManager::CommonUpdate<BehaviourManager>())
  #33  pc 0x00000000001614eb  /data/app/com.fourxp.mergeup-VlNZ4_OQ3Q7uuUnb5kDQNA==/lib/arm/libunity.so (ExecutePlayerLoop(NativePlayerLoopSystem*))
  #34  pc 0x0000000000161507  /data/app/com.fourxp.mergeup-VlNZ4_OQ3Q7uuUnb5kDQNA==/lib/arm/libunity.so (ExecutePlayerLoop(NativePlayerLoopSystem*))
  #35  pc 0x00000000001616a1  /data/app/com.fourxp.mergeup-VlNZ4_OQ3Q7uuUnb5kDQNA==/lib/arm/libunity.so (PlayerLoop())
  #36  pc 0x0000000000229e21  /data/app/com.fourxp.mergeup-VlNZ4_OQ3Q7uuUnb5kDQNA==/lib/arm/libunity.so (UnityPlayerLoop())
  #37  pc 0x00000000002396cd  /data/app/com.fourxp.mergeup-VlNZ4_OQ3Q7uuUnb5kDQNA==/lib/arm/libunity.so (nativeRender(_JNIEnv*, _jobject*))
  #38  pc 0x0000000000041033  /data/app/com.fourxp.mergeup-VlNZ4_OQ3Q7uuUnb5kDQNA==/oat/arm/base.odex

<img width="1308" alt="Screen Shot 2023-01-03 at 8 22 25 AM" src="https://user-images.githubusercontent.com/1131364/210308710-829037ab-bacf-4f48-a766-82e2d4955b9f.png">
*** *** *** *** *** *** *** *** *** *** *** *** *** *** *** ***
pid: 0, tid: 11499 >>> com.fourxp.mergeup <<<

backtrace:
  #00  pc 0x0000000000078b9c  /apex/com.android.runtime/lib64/bionic/libc.so (abort+180)
  #01  pc 0x00000000006d157c  /apex/com.android.art/lib64/libart.so (art::Runtime::Abort(char const*)+668)
  #02  pc 0x000000000001695c  /apex/com.android.art/lib64/libbase.so
  #03  pc 0x0000000000015f8c  /apex/com.android.art/lib64/libbase.so (android::base::LogMessage::~LogMessage()+364)
  #04  pc 0x00000000002857ec  /apex/com.android.art/lib64/libart.so (art::JavaVMExt::AddGlobalRef(art::Thread*, art::ObjPtr<art::mirror::Object>)+1924)
  #05  pc 0x000000000059add0  /apex/com.android.art/lib64/libart.so (art::JNI<false>::NewGlobalRef(_JNIEnv*, _jobject*)+272)
  #06  pc 0x0000000000251d0c  /data/app/~~EtwUZTc56Y3a30rY1f1XwQ==/com.fourxp.mergeup-FN-r4MexkF1MsOt-2IZHZQ==/lib/arm64/libFirebaseCppApp-10_1_1.so (firebase::firestore::FieldValueInternal::FieldValueInternal(long)+176)
  #07  pc 0x0000000000240c1c  /data/app/~~EtwUZTc56Y3a30rY1f1XwQ==/com.fourxp.mergeup-FN-r4MexkF1MsOt-2IZHZQ==/lib/arm64/libFirebaseCppApp-10_1_1.so (firebase::firestore::FieldValue::Integer(long)+40)
  #08  pc 0x00000000001f1320  /data/app/~~EtwUZTc56Y3a30rY1f1XwQ==/com.fourxp.mergeup-FN-r4MexkF1MsOt-2IZHZQ==/lib/arm64/libFirebaseCppApp-10_1_1.so (Firebase_Firestore_CSharp_FieldValueProxy_Integer+48)
  #09  pc 0x00000000005b32b8  /data/app/~~EtwUZTc56Y3a30rY1f1XwQ==/com.fourxp.mergeup-FN-r4MexkF1MsOt-2IZHZQ==/lib/arm64/libil2cpp.so (FirestoreCppPINVOKE_FieldValueProxy_Integer_mFA22725C6FCB7674F6EFC851553D99C0B998346E+28520)
  #10  pc 0x00000000005a5e64  /data/app/~~EtwUZTc56Y3a30rY1f1XwQ==/com.fourxp.mergeup-FN-r4MexkF1MsOt-2IZHZQ==/lib/arm64/libil2cpp.so (FieldValueProxy_Integer_m347744588961CEF34B8D91B7A7F75E4E433F4EB7+23713)
  #11  pc 0x00000000005b11f0  /data/app/~~EtwUZTc56Y3a30rY1f1XwQ==/com.fourxp.mergeup-FN-r4MexkF1MsOt-2IZHZQ==/lib/arm64/libil2cpp.so (EnumerableConverterBase_Serialize_m40FF859EA17BAEE65D7765ADF75C903F1DE77792+21357)
  #12  pc 0x00000000005a4e74  /data/app/~~EtwUZTc56Y3a30rY1f1XwQ==/com.fourxp.mergeup-FN-r4MexkF1MsOt-2IZHZQ==/lib/arm64/libil2cpp.so (AttributedTypeConverter_SerializeMap_m2AA6C554C645BF93A347C40E5D0F00F43FC8C10B+14823)
  #13  pc 0x00000000005bf4bc  /data/app/~~EtwUZTc56Y3a30rY1f1XwQ==/com.fourxp.mergeup-FN-r4MexkF1MsOt-2IZHZQ==/lib/arm64/libil2cpp.so (MapConverterBase_Serialize_mF24CAD26C3A877991AE6732907C94C39078FD206+44)
  #14  pc 0x00000000005a6604  /data/app/~~EtwUZTc56Y3a30rY1f1XwQ==/com.fourxp.mergeup-FN-r4MexkF1MsOt-2IZHZQ==/lib/arm64/libil2cpp.so (DocumentReference_SetAsync_mA0B8D6AB674F0ADF47F693B58D5309B8B84E7451+18859)
  #15  pc 0x0000000000ae6c74  /data/app/~~EtwUZTc56Y3a30rY1f1XwQ==/com.fourxp.mergeup-FN-r4MexkF1MsOt-2IZHZQ==/lib/arm64/libil2cpp.so (FirebaseManager_SendUserDataToFireStore_mC7C7CA7F6A5105EEE48ABEB25BA7F4671C99DCEC+39600)
  #16  pc 0x0000000000af7c60  /data/app/~~EtwUZTc56Y3a30rY1f1XwQ==/com.fourxp.mergeup-FN-r4MexkF1MsOt-2IZHZQ==/lib/arm64/libil2cpp.so (Game_saveGameState_m7A716F3F5B2506DD1ABF74D4400B435BC62429FF+31023)
  #17  pc 0x0000000000afb898  /data/app/~~EtwUZTc56Y3a30rY1f1XwQ==/com.fourxp.mergeup-FN-r4MexkF1MsOt-2IZHZQ==/lib/arm64/libil2cpp.so (Game_Update_m3BB58C19CD85458B683F7137A83EE9EA61700396+31414)
  #18  pc 0x000000000029b7bc  /data/app/~~EtwUZTc56Y3a30rY1f1XwQ==/com.fourxp.mergeup-FN-r4MexkF1MsOt-2IZHZQ==/lib/arm64/libil2cpp.so (RuntimeInvoker_TrueVoid_t700C6383A2A510C2CF4DD86DABD5CA9FF70ADAC5(void (*)(), MethodInfo const*, void*, void**)+44228)
  #19  pc 0x000000000035d4b8  /data/app/~~EtwUZTc56Y3a30rY1f1XwQ==/com.fourxp.mergeup-FN-r4MexkF1MsOt-2IZHZQ==/lib/arm64/libil2cpp.so (il2cpp::vm::Runtime::Invoke(MethodInfo const*, void*, void**, Il2CppException**)+556)
  #20  pc 0x00000000002bebc0  /data/app/~~EtwUZTc56Y3a30rY1f1XwQ==/com.fourxp.mergeup-FN-r4MexkF1MsOt-2IZHZQ==/lib/arm64/libunity.so (scripting_method_invoke(ScriptingMethodPtr, ScriptingObjectPtr, ScriptingArguments&, ScriptingExceptionPtr*, bool))
  #21  pc 0x00000000002c21fc  /data/app/~~EtwUZTc56Y3a30rY1f1XwQ==/com.fourxp.mergeup-FN-r4MexkF1MsOt-2IZHZQ==/lib/arm64/libunity.so (ScriptingInvocation::Invoke(ScriptingExceptionPtr*, bool))
  #22  pc 0x00000000002d76a4  /data/app/~~EtwUZTc56Y3a30rY1f1XwQ==/com.fourxp.mergeup-FN-r4MexkF1MsOt-2IZHZQ==/lib/arm64/libunity.so (MonoBehaviour::CallUpdateMethod(int))
  #23  pc 0x00000000001ad0cc  /data/app/~~EtwUZTc56Y3a30rY1f1XwQ==/com.fourxp.mergeup-FN-r4MexkF1MsOt-2IZHZQ==/lib/arm64/libunity.so (void BaseBehaviourManager::CommonUpdate<BehaviourManager>())
  #24  pc 0x000000000022456c  /data/app/~~EtwUZTc56Y3a30rY1f1XwQ==/com.fourxp.mergeup-FN-r4MexkF1MsOt-2IZHZQ==/lib/arm64/libunity.so (ExecutePlayerLoop(NativePlayerLoopSystem*))
  #25  pc 0x00000000002245a0  /data/app/~~EtwUZTc56Y3a30rY1f1XwQ==/com.fourxp.mergeup-FN-r4MexkF1MsOt-2IZHZQ==/lib/arm64/libunity.so (ExecutePlayerLoop(NativePlayerLoopSystem*))
  #26  pc 0x00000000002247e4  /data/app/~~EtwUZTc56Y3a30rY1f1XwQ==/com.fourxp.mergeup-FN-r4MexkF1MsOt-2IZHZQ==/lib/arm64/libunity.so (PlayerLoop())
  #27  pc 0x000000000033b608  /data/app/~~EtwUZTc56Y3a30rY1f1XwQ==/com.fourxp.mergeup-FN-r4MexkF1MsOt-2IZHZQ==/lib/arm64/libunity.so (UnityPlayerLoop())
  #28  pc 0x0000000000351158  /data/app/~~EtwUZTc56Y3a30rY1f1XwQ==/com.fourxp.mergeup-FN-r4MexkF1MsOt-2IZHZQ==/lib/arm64/libunity.so (nativeRender(_JNIEnv*, _jobject*))
  #29  pc 0x000000000200394c  /memfd:jit-cache

<img width="1316" alt="Screen Shot 2023-01-03 at 8 22 14 AM" src="https://user-images.githubusercontent.com/1131364/210308715-619a7953-7b76-4cf9-9981-8537fbe79315.png">
*** *** *** *** *** *** *** *** *** *** *** *** *** *** *** ***
pid: 0, tid: 16730 >>> com.fourxp.mergeup <<<

backtrace:
  #00  pc 0x000000000005f2b6  /apex/com.android.runtime/lib/bionic/libc.so (abort+166)
  #01  pc 0x0000000000376789  /apex/com.android.runtime/lib/libart.so (art::Runtime::Abort(char const*)+1676)
  #02  pc 0x000000000000855f  /system/lib/libbase.so (android::base::LogMessage::~LogMessage()+406)
  #03  pc 0x0000000000288c79  /apex/com.android.runtime/lib/libart.so (art::JavaVMExt::AddGlobalRef(art::Thread*, art::ObjPtr<art::mirror::Object>)+272)
  #04  pc 0x0000000000292c1d  /apex/com.android.runtime/lib/libart.so (art::JNI::NewGlobalRef(_JNIEnv*, _jobject*)+428)
  #05  pc 0x000000000021ea09  /data/app/com.fourxp.mergeup-7f7CWfmf6ngQ0NX3SfqL1w==/lib/arm/libFirebaseCppApp-10_1_1.so (firebase::firestore::jni::Global<firebase::firestore::jni::Object>::operator=(firebase::firestore::jni::Object const&)+58)
  #06  pc 0x000000000021ea89  /data/app/com.fourxp.mergeup-7f7CWfmf6ngQ0NX3SfqL1w==/lib/arm/libFirebaseCppApp-10_1_1.so (firebase::firestore::FieldValueInternal::FieldValueInternal(long long)+72)
  #07  pc 0x0000000000214541  /data/app/com.fourxp.mergeup-7f7CWfmf6ngQ0NX3SfqL1w==/lib/arm/libFirebaseCppApp-10_1_1.so (firebase::firestore::FieldValue::Integer(long long)+26)
  #08  pc 0x00000000001e4dad  /data/app/com.fourxp.mergeup-7f7CWfmf6ngQ0NX3SfqL1w==/lib/arm/libFirebaseCppApp-10_1_1.so (Firebase_Firestore_CSharp_FieldValueProxy_Integer+36)
  #09  pc 0x000000000047c5ac  /data/app/com.fourxp.mergeup-7f7CWfmf6ngQ0NX3SfqL1w==/lib/arm/libil2cpp.so (FirestoreCppPINVOKE_FieldValueProxy_Integer_mFA22725C6FCB7674F6EFC851553D99C0B998346E+28520)
  #10  pc 0x000000000046c764  /data/app/com.fourxp.mergeup-7f7CWfmf6ngQ0NX3SfqL1w==/lib/arm/libil2cpp.so (FieldValueProxy_Integer_m347744588961CEF34B8D91B7A7F75E4E433F4EB7+23713)
  #11  pc 0x0000000000479f50  /data/app/com.fourxp.mergeup-7f7CWfmf6ngQ0NX3SfqL1w==/lib/arm/libil2cpp.so (EnumerableConverterBase_Serialize_m40FF859EA17BAEE65D7765ADF75C903F1DE77792+21357)
  #12  pc 0x000000000046b488  /data/app/com.fourxp.mergeup-7f7CWfmf6ngQ0NX3SfqL1w==/lib/arm/libil2cpp.so (AttributedTypeConverter_SerializeMap_m2AA6C554C645BF93A347C40E5D0F00F43FC8C10B+14823)
  #13  pc 0x000000000048aeec  /data/app/com.fourxp.mergeup-7f7CWfmf6ngQ0NX3SfqL1w==/lib/arm/libil2cpp.so (MapConverterBase_Serialize_mF24CAD26C3A877991AE6732907C94C39078FD206+44)
  #14  pc 0x000000000046d020  /data/app/com.fourxp.mergeup-7f7CWfmf6ngQ0NX3SfqL1w==/lib/arm/libil2cpp.so (DocumentReference_SetAsync_mA0B8D6AB674F0ADF47F693B58D5309B8B84E7451+18859)
  #15  pc 0x0000000000ab0428  /data/app/com.fourxp.mergeup-7f7CWfmf6ngQ0NX3SfqL1w==/lib/arm/libil2cpp.so (FirebaseManager_SendBusinessDataToFireStore_m6452EF303E7E3AA9A6FA43904C0B5084528B73A3+39835)
  #16  pc 0x0000000000ac8224  /data/app/com.fourxp.mergeup-7f7CWfmf6ngQ0NX3SfqL1w==/lib/arm/libil2cpp.so (Game_SaveBusinessData_mB4107BD702082927FDD9FEF420BDEC0AB8A33114+31062)
  #17  pc 0x0000000000705054  /data/app/com.fourxp.mergeup-7f7CWfmf6ngQ0NX3SfqL1w==/lib/arm/libil2cpp.so (Action_1_Invoke_m587509C88BB83721D7918D89DF07606BB752D744_gshared+12812)
  #18  pc 0x0000000000a7df24  /data/app/com.fourxp.mergeup-7f7CWfmf6ngQ0NX3SfqL1w==/lib/arm/libil2cpp.so (U3CU3Ec__DisplayClass0_1_U3CContinueWithOnMainThreadU3Eb__1_m669E73DC507424634BD00848B6A61F918DCF75DA+921)
  #19  pc 0x00000000006d517c  /data/app/com.fourxp.mergeup-7f7CWfmf6ngQ0NX3SfqL1w==/lib/arm/libil2cpp.so (Func_1_Invoke_mD0192F3F912E371387CD7B196DF85569553486B4_gshared+19691)
  #20  pc 0x000000000070102c  /data/app/com.fourxp.mergeup-7f7CWfmf6ngQ0NX3SfqL1w==/lib/arm/libil2cpp.so (U3CU3Ec__DisplayClass5_0_1_U3CRunAsyncU3Eb__0_m0AAB5C27727D26D190EB040F6E10869615D38C75_gshared+9925)
  #21  pc 0x000000000065efc8  /data/app/com.fourxp.mergeup-7f7CWfmf6ngQ0NX3SfqL1w==/lib/arm/libil2cpp.so (Action_Invoke_m3FFA5BE3D64F0FF8E1E1CB6F953913FADB5EB89E+21048)
  #22  pc 0x0000000000a4fe18  /data/app/com.fourxp.mergeup-7f7CWfmf6ngQ0NX3SfqL1w==/lib/arm/libil2cpp.so (ExceptionAggregator_Wrap_m286D2940E6E4C43E3C9B5E23A249A63B731390B4+5296)
  #23  pc 0x0000000000a4fcd0  /data/app/com.fourxp.mergeup-7f7CWfmf6ngQ0NX3SfqL1w==/lib/arm/libil2cpp.so (Dispatcher_PollJobs_m4DD47E8B63F33D975BC1A7D670A509F597F8E9E0+4915)
  #24  pc 0x0000000000a5152c  /data/app/com.fourxp.mergeup-7f7CWfmf6ngQ0NX3SfqL1w==/lib/arm/libil2cpp.so (FirebaseHandler_Update_m054C867D529EEC80853232EEAFBCAD18CC287817+6523)
  #25  pc 0x00000000001236e4  /data/app/com.fourxp.mergeup-7f7CWfmf6ngQ0NX3SfqL1w==/lib/arm/libil2cpp.so (RuntimeInvoker_TrueVoid_t700C6383A2A510C2CF4DD86DABD5CA9FF70ADAC5(void (*)(), MethodInfo const*, void*, void**)+44228)
  #26  pc 0x00000000001ec848  /data/app/com.fourxp.mergeup-7f7CWfmf6ngQ0NX3SfqL1w==/lib/arm/libil2cpp.so (il2cpp::vm::Runtime::Invoke(MethodInfo const*, void*, void**, Il2CppException**)+556)
  #27  pc 0x00000000001d23a1  /data/app/com.fourxp.mergeup-7f7CWfmf6ngQ0NX3SfqL1w==/lib/arm/libunity.so (scripting_method_invoke(ScriptingMethodPtr, ScriptingObjectPtr, ScriptingArguments&, ScriptingExceptionPtr*, bool))
  #28  pc 0x00000000001d4aa1  /data/app/com.fourxp.mergeup-7f7CWfmf6ngQ0NX3SfqL1w==/lib/arm/libunity.so (ScriptingInvocation::Invoke(ScriptingExceptionPtr*, bool))
  #29  pc 0x00000000001e43e1  /data/app/com.fourxp.mergeup-7f7CWfmf6ngQ0NX3SfqL1w==/lib/arm/libunity.so (MonoBehaviour::CallUpdateMethod(int))
  #30  pc 0x0000000000108c1f  /data/app/com.fourxp.mergeup-7f7CWfmf6ngQ0NX3SfqL1w==/lib/arm/libunity.so (void BaseBehaviourManager::CommonUpdate<BehaviourManager>())
  #31  pc 0x00000000001614eb  /data/app/com.fourxp.mergeup-7f7CWfmf6ngQ0NX3SfqL1w==/lib/arm/libunity.so (ExecutePlayerLoop(NativePlayerLoopSystem*))
  #32  pc 0x0000000000161507  /data/app/com.fourxp.mergeup-7f7CWfmf6ngQ0NX3SfqL1w==/lib/arm/libunity.so (ExecutePlayerLoop(NativePlayerLoopSystem*))
  #33  pc 0x00000000001616a1  /data/app/com.fourxp.mergeup-7f7CWfmf6ngQ0NX3SfqL1w==/lib/arm/libunity.so (PlayerLoop())
  #34  pc 0x0000000000229e21  /data/app/com.fourxp.mergeup-7f7CWfmf6ngQ0NX3SfqL1w==/lib/arm/libunity.so (UnityPlayerLoop())
  #35  pc 0x00000000002396cd  /data/app/com.fourxp.mergeup-7f7CWfmf6ngQ0NX3SfqL1w==/lib/arm/libunity.so (nativeRender(_JNIEnv*, _jobject*))
  #36  pc 0x0000000000041033  /data/app/com.fourxp.mergeup-7f7CWfmf6ngQ0NX3SfqL1w==/oat/arm/base.odex

<img width="1320" alt="Screen Shot 2023-01-03 at 8 22 04 AM" src="https://user-images.githubusercontent.com/1131364/210308718-966780f4-7f82-472a-bed2-5804d3e8e3f9.png">
*** *** *** *** *** *** *** *** *** *** *** *** *** *** *** ***
pid: 0, tid: 17430 >>> com.fourxp.mergeup <<<

backtrace:
  #00  pc 0x000000000004fe7c  /apex/com.android.runtime/lib64/bionic/libc.so (abort+180)
  #01  pc 0x00000000006d122c  /apex/com.android.art/lib64/libart.so (art::Runtime::Abort(char const*)+704)
  #02  pc 0x0000000000016ea8  /apex/com.android.art/lib64/libbase.so
  #03  pc 0x0000000000016450  /apex/com.android.art/lib64/libbase.so (android::base::LogMessage::~LogMessage()+352)
  #04  pc 0x000000000045cacc  /apex/com.android.art/lib64/libart.so (art::JavaVMExt::AddGlobalRef(art::Thread*, art::ObjPtr<art::mirror::Object>)+1908)
  #05  pc 0x00000000005cf9b4  /apex/com.android.art/lib64/libart.so (art::JNI<false>::NewGlobalRef(_JNIEnv*, _jobject*)+152)
  #06  pc 0x00000000002407f4  /data/app/~~UooMqfqalDpYG3uZsCxRWQ==/com.fourxp.mergeup-1E4bQk7m0CpDAXWTYN7KLQ==/lib/arm64/libFirebaseCppApp-10_1_1.so (firebase::firestore::FieldValue::FieldValue(firebase::firestore::FieldValue const&)+100)
  #07  pc 0x000000000020830c  /data/app/~~UooMqfqalDpYG3uZsCxRWQ==/com.fourxp.mergeup-1E4bQk7m0CpDAXWTYN7KLQ==/lib/arm64/libFirebaseCppApp-10_1_1.so (Firebase_Firestore_CSharp_FieldValueVector_PushBack+40)
  #08  pc 0x00000000005b4bcc  /data/app/~~UooMqfqalDpYG3uZsCxRWQ==/com.fourxp.mergeup-1E4bQk7m0CpDAXWTYN7KLQ==/lib/arm64/libil2cpp.so (FirestoreCppPINVOKE_FieldValueVector_PushBack_m6F53C48070883074B9BC581A38373E5C2A1C4F5B+30945)
  #09  pc 0x00000000005a8dbc  /data/app/~~UooMqfqalDpYG3uZsCxRWQ==/com.fourxp.mergeup-1E4bQk7m0CpDAXWTYN7KLQ==/lib/arm64/libil2cpp.so (FieldValueVector_PushBack_mA57371306F100708C468FA543AF22A302D4B0E64+25282)
  #10  pc 0x00000000005a8ba4  /data/app/~~UooMqfqalDpYG3uZsCxRWQ==/com.fourxp.mergeup-1E4bQk7m0CpDAXWTYN7KLQ==/lib/arm64/libil2cpp.so (ConverterBase_ConvertToProxyArray_m15B85B6AC3B0D8B4B7D9B84D516844EF3FF1E4C4+16730)
  #11  pc 0x00000000005b12c0  /data/app/~~UooMqfqalDpYG3uZsCxRWQ==/com.fourxp.mergeup-1E4bQk7m0CpDAXWTYN7KLQ==/lib/arm64/libil2cpp.so (EnumerableConverterBase_Serialize_m40FF859EA17BAEE65D7765ADF75C903F1DE77792+21422)
  #12  pc 0x00000000005a4e74  /data/app/~~UooMqfqalDpYG3uZsCxRWQ==/com.fourxp.mergeup-1E4bQk7m0CpDAXWTYN7KLQ==/lib/arm64/libil2cpp.so (AttributedTypeConverter_SerializeMap_m2AA6C554C645BF93A347C40E5D0F00F43FC8C10B+14823)
  #13  pc 0x00000000005bf4bc  /data/app/~~UooMqfqalDpYG3uZsCxRWQ==/com.fourxp.mergeup-1E4bQk7m0CpDAXWTYN7KLQ==/lib/arm64/libil2cpp.so (MapConverterBase_Serialize_mF24CAD26C3A877991AE6732907C94C39078FD206+44)
  #14  pc 0x00000000005a6604  /data/app/~~UooMqfqalDpYG3uZsCxRWQ==/com.fourxp.mergeup-1E4bQk7m0CpDAXWTYN7KLQ==/lib/arm64/libil2cpp.so (DocumentReference_SetAsync_mA0B8D6AB674F0ADF47F693B58D5309B8B84E7451+18859)
  #15  pc 0x0000000000ae6c74  /data/app/~~UooMqfqalDpYG3uZsCxRWQ==/com.fourxp.mergeup-1E4bQk7m0CpDAXWTYN7KLQ==/lib/arm64/libil2cpp.so (FirebaseManager_SendUserDataToFireStore_mC7C7CA7F6A5105EEE48ABEB25BA7F4671C99DCEC+39600)
  #16  pc 0x0000000000af7c60  /data/app/~~UooMqfqalDpYG3uZsCxRWQ==/com.fourxp.mergeup-1E4bQk7m0CpDAXWTYN7KLQ==/lib/arm64/libil2cpp.so (Game_saveGameState_m7A716F3F5B2506DD1ABF74D4400B435BC62429FF+31023)
  #17  pc 0x0000000000afb898  /data/app/~~UooMqfqalDpYG3uZsCxRWQ==/com.fourxp.mergeup-1E4bQk7m0CpDAXWTYN7KLQ==/lib/arm64/libil2cpp.so (Game_Update_m3BB58C19CD85458B683F7137A83EE9EA61700396+31414)
  #18  pc 0x000000000029b7bc  /data/app/~~UooMqfqalDpYG3uZsCxRWQ==/com.fourxp.mergeup-1E4bQk7m0CpDAXWTYN7KLQ==/lib/arm64/libil2cpp.so (RuntimeInvoker_TrueVoid_t700C6383A2A510C2CF4DD86DABD5CA9FF70ADAC5(void (*)(), MethodInfo const*, void*, void**)+44228)
  #19  pc 0x000000000035d4b8  /data/app/~~UooMqfqalDpYG3uZsCxRWQ==/com.fourxp.mergeup-1E4bQk7m0CpDAXWTYN7KLQ==/lib/arm64/libil2cpp.so (il2cpp::vm::Runtime::Invoke(MethodInfo const*, void*, void**, Il2CppException**)+556)
  #20  pc 0x00000000002bebc0  /data/app/~~UooMqfqalDpYG3uZsCxRWQ==/com.fourxp.mergeup-1E4bQk7m0CpDAXWTYN7KLQ==/lib/arm64/libunity.so (scripting_method_invoke(ScriptingMethodPtr, ScriptingObjectPtr, ScriptingArguments&, ScriptingExceptionPtr*, bool))
  #21  pc 0x00000000002c21fc  /data/app/~~UooMqfqalDpYG3uZsCxRWQ==/com.fourxp.mergeup-1E4bQk7m0CpDAXWTYN7KLQ==/lib/arm64/libunity.so (ScriptingInvocation::Invoke(ScriptingExceptionPtr*, bool))
  #22  pc 0x00000000002d76a4  /data/app/~~UooMqfqalDpYG3uZsCxRWQ==/com.fourxp.mergeup-1E4bQk7m0CpDAXWTYN7KLQ==/lib/arm64/libunity.so (MonoBehaviour::CallUpdateMethod(int))
  #23  pc 0x00000000001ad0cc  /data/app/~~UooMqfqalDpYG3uZsCxRWQ==/com.fourxp.mergeup-1E4bQk7m0CpDAXWTYN7KLQ==/lib/arm64/libunity.so (void BaseBehaviourManager::CommonUpdate<BehaviourManager>())
  #24  pc 0x000000000022456c  /data/app/~~UooMqfqalDpYG3uZsCxRWQ==/com.fourxp.mergeup-1E4bQk7m0CpDAXWTYN7KLQ==/lib/arm64/libunity.so (ExecutePlayerLoop(NativePlayerLoopSystem*))
  #25  pc 0x00000000002245a0  /data/app/~~UooMqfqalDpYG3uZsCxRWQ==/com.fourxp.mergeup-1E4bQk7m0CpDAXWTYN7KLQ==/lib/arm64/libunity.so (ExecutePlayerLoop(NativePlayerLoopSystem*))
  #26  pc 0x00000000002247e4  /data/app/~~UooMqfqalDpYG3uZsCxRWQ==/com.fourxp.mergeup-1E4bQk7m0CpDAXWTYN7KLQ==/lib/arm64/libunity.so (PlayerLoop())
  #27  pc 0x000000000033b608  /data/app/~~UooMqfqalDpYG3uZsCxRWQ==/com.fourxp.mergeup-1E4bQk7m0CpDAXWTYN7KLQ==/lib/arm64/libunity.so (UnityPlayerLoop())
  #28  pc 0x0000000000351158  /data/app/~~UooMqfqalDpYG3uZsCxRWQ==/com.fourxp.mergeup-1E4bQk7m0CpDAXWTYN7KLQ==/lib/arm64/libunity.so (nativeRender(_JNIEnv*, _jobject*))
  #29  pc 0x000000000043dd54  /apex/com.android.art/lib64/libart.so (art_quick_generic_jni_trampoline+148)
  #30  pc 0x000000000020a2b0  /apex/com.android.art/lib64/libart.so (nterp_helper+4016)
  #31  pc 0x0000000002002694  /memfd:jit-cache

**Comment by nitzanwilnai** [comment]
We also believe that the non-firebase issues on Google Play are also related:
<img width="1351" alt="Screen Shot 2023-01-03 at 8 33 16 AM" src="https://user-images.githubusercontent.com/1131364/210309729-b3fba3e3-d3dc-4768-8e0b-d836aa9c0832.png">

For example the various `[libc.so]` abort and `[libc.so] prop_area::map_prop_area_rw(char const*, char const*, bool*)` errors.

But none of them have any stack trace.

Note also that we have not been able to reproduce any of these issues locally, they only happen in production.

**Comment by nitzanwilnai** [comment]
Note that updating to Firebase 10.3.0 did not fix the issue.

**Comment by nitzanwilnai** [comment]
So our documents are fairly optimized (bytes instead of ints, single letter variable names, etc...)
We are sending the same document fairly often. (the game data)
We are also sending one way - we receive the game data at loading, afterwards we only send the new data.
Is there a way to manually clear the pool?

**Comment by dconeybe** [comment]
I've looked at your crash reports more closely and can confirm 100% that you are experiencing the Android "global refs exhaustion issue" (https://github.com/firebase/firebase-unity-sdk/issues/569). The fact that the crashes are occurring in `art::JNI::NewGlobalRef(...)` confirms this.

Now, it's not the size of the documents, in bytes, that matters, but rather the number of key value pairs and the number of elements in array and map values. Each value and each array/map element consumes a JNI global reference, along with each Firestore object (e.g. DocumentSnapshot, Transaction). Unfortunately, there really is no workaround other than reducing the number of Firestore objects in memory at a given time. If you have documents with large array or map values, then avoid keeping those documents in memory.

The upcoming fix will likely address this crash though. It will, unfortunately, be a month until the release of the fix though.

I'm going to close this issue since it is a duplicate of https://github.com/firebase/firebase-unity-sdk/issues/569. Please follow that issue for updates. We can still discuss here in this issue if you have follow-up quetsions.

**Comment by nitzanwilnai** [comment]
Can you explain to me a bit more what you mean by keeping these documents in memory?
Are you talking about the memory during the SetAsync call? When is this memory freed?

Or are you talking about the copy of my [FireStore Data] class?

Thanks!

**Comment by dconeybe** [comment]
Oh, all I mean is holding on to DocumentSnapshot objects. Make sure you null out references to Firestore objects once you don't need them anymore (or let them go out of scope so they will be garbage collected). If your app is well-written, there may not be any opportunity here to do that though.

**Comment by nitzanwilnai** [comment]
So in my SendUserDataToFireStore code here:
`        public void SendUserDataToFireStore(UserData userData)
        {
            FirebaseTimeout = Constants.FIREBASE_TIMEOUT;

            FirebaseFirestore db = FirebaseFirestore.DefaultInstance;

            if (PlayerPrefs.GetInt(FirebaseUser.UserId) == 0)
                PlayerPrefs.SetInt(FirebaseUser.UserId, 1);

            DocumentReference docRef = db.Collection(Constants.UserDataCollection).Document(FirebaseUser.UserId);
            docRef.SetAsync(userData).ContinueWith(task =>
            {
                if (task.IsCompleted)
                {

                    Debug.Log("UserData DocRef ID " + docRef.Id);

                }
                else
                {
                    Debug.Log("SendUserDataToFireStore ERROR ADDING DATA TO FIRESTORE!");
                }
            });
        }`

Do I need to save a reference to the docRef, then set it to null once the ContinueWith returns?

**Comment by dconeybe** [comment]
The DocumentReference only consumes 1 global reference. So unless you have a LOT of DocumentReference objects alive at once then it doesn't matter. But to answer your question, you should set docRef to null ASAP, unless it just goes out of scope, which has the same effect as setting it to null. So probably the cruft of setting it to null is not worth it.

The more interesting object is the `userData` object. How many key/value pairs does it contain? Does it contain any array or map values?

**Comment by nitzanwilnai** [comment]
So yes, we have two big objects, userData and businessData, and they both have a few arrays, some of them quite large. In one we had an array with 1024 integers.

We also have nested arrays of other FirestoreData classes. 
For example we have our BusinessData class, which references BoardData, which has arrays of ProducerData, AutoMorphData, BlenderData, etc...

We have tried to reduce everything to as small as possible and so far are seeing fewer crashes.

That said, I still don't understand why the crashes happen after 20+ minutes of gameplay and dozens of SetAsync calls if the memory gets freed once the SetAsync call is done.

**Comment by nitzanwilnai** [comment]
Here is our userData class:
`
using System;
using System.Collections;
using System.Collections.Generic;
using System.Runtime.InteropServices;
using Firebase.Firestore;
using UnityEngine;

namespace MergeGame
{
    [FirestoreData]
    public class UserData
    {
        [FirestoreProperty] public int V { get; set; } // version
        [FirestoreProperty] public string AB { get; set; } // AB Test

        [FirestoreProperty] public DateTime T1 { get; set; } // DateTime of First login event
        [FirestoreProperty] public DateTime TC { get; set; } // DateTime of Last login event sent to analytics
        [FirestoreProperty] public int G { get; set; } // Group, from 1-12. Will be used for A/B testing in the future
        [FirestoreProperty] public string N { get; set; } // username
        [FirestoreProperty] public bool A { get; set; } // admin
        [FirestoreProperty] public int Gems { get; set; }
        [FirestoreProperty] public short Energy { get; set; }
        [FirestoreProperty] public long EnergyTime { get; set; }

        [FirestoreProperty] public int[] DiscoveredFullIDs { get; set; }
        [FirestoreProperty] public short DiscoveredCount { get; set; }

        [FirestoreProperty] public short[] MiniTutorialChain { get; set; }
        [FirestoreProperty] public short MiniTutorialChainCount { get; set; }

        [FirestoreProperty] public int[] MiniTutorialItem { get; set; }
        [FirestoreProperty] public short MiniTutorialItemCount { get; set; }

        [FirestoreProperty] public bool MiniTutorialProducerRefill { get; set; }
        [FirestoreProperty] public bool MiniTutorialEnergyRefill { get; set; }
        [FirestoreProperty] public bool MTB { get; set; } // Mini tutorial bubble

        [FirestoreProperty] public int DebugBundleCounter { get; set; }

        [FirestoreProperty] public bool MuteSFX { get; set; }
        [FirestoreProperty] public bool MuteMusic { get; set; }

        [FirestoreProperty] public bool DisableLocalNotifications { get; set; }

        [FirestoreProperty] public bool UserRatedApp { get; set; }
        [FirestoreProperty] public int RateUsPopupLastOpenedLevel { get; set; }
    }

}
`

You can tell that at some point we actually read the Firestore doc and started changing our names to the shortest possible. ;)
I assume name length does not affect this issue, only the size of the data stored?

**Comment by dconeybe** [comment]
Yep, those objects look like heavy global ref consumers. Probably what is happening is that the Firestore C# objects are accumulating in memory, even though they are no longer referenced. Eventually, the garbage collector runs, destroys the C# objects, and those objects' finalizers are run to free the Android global refs. But if the number of global refs hits the 51200 hard maximum before the garbage collector runs then the exhaustion occurs, resulting in the crash. That's probably why it takes 20 minutes or so before the global refs accumulate to a problematic level.

The _only_ workaround that I can think of would be to call `FirebaseApp.Dispose()` periodically to free some global refs. This is, of course, a very heavy operation and you will need to re-create the FirebaseApp and the FirebaseFirestore, etc. again from scratch. But I'm throwing it out there as a potential workaround.

**Comment by nitzanwilnai** [comment]
Is Firebase using its own garbage collector or is it using Unity's?

Can I manually cache and re-use the document refs instead of creating new ones?

**Comment by dconeybe** [comment]
Firebase is _not_ using its own garbage collector; it's using Unity's.

No, there is no way to manually re-use the document refs instead of creating new ones?

**Comment by nitzanwilnai** [comment]
In that case, would calling System.GC.Collect() or Resources.UnloadUnusedAssets manually at opportune times solve the issue?

**Comment by dconeybe** [comment]
It's worth a try, although I don't know for sure. It depends how the garbage collector responds to those calls.

**Comment by nitzanwilnai** [comment]
FYI, unfortunately calling:
                Resources.UnloadUnusedAssets();
                GC.Collect();
Does not solve the problem.

**Comment by nitzanwilnai** [comment]
The only solution we found is:
1. reducing the number of references in our FirestoreData classes
2. reducing the number of times we save to Firestore.

At least on some devices, there doesn't seem to be any clearing of the reference pool, ever, during gameplay. Possibly on all devices.

To me it looks more and more like you guys have a memory leak in the code.

**Comment by dconeybe** [comment]
Interesting. It's entirely possible that there is a memory leak in Firestore. Would you be able to use something like Android's "Memory Profiler" to see if you can get some more details about the memory usage of your app, and/or find some evidence of the unbounded memory growth? This information could point to the part of the Firestore SDK that is leaking memory.

https://developer.android.com/studio/profile/memory-profiler

**Comment by nitzanwilnai** [comment]
Hi, I do not see any sort of leak in the memory profiler. 
That said, I don't think we would see one. The issue is not new memory being allocated, its that the reference pool is not clearing itself correctly when the docRef is garbage collected. 
I assume the pool would be pre-allocated to 51200 elements, so adding to the pool wouldn't show up in the memory profiler anyways.

**Comment by nitzanwilnai** [comment]
Hi, another question. You mentioned an array of X elements would be X+1 references (if I understood you correctly.)
What about a string? Would a string be 1 reference or would it be treated like an array of bytes?

**Comment by dconeybe** [comment]
A String is just 1 global reference (the reference to the java.lang.String Java object).
